tbot/machine/linux/path.py

```diff
@@ -71,13 +61,152 @@ def __init__(self, host: H, *args: typing.Any) -> None:
+    # These methods are tbot-specific and do not have an equivalent in pathlib.
 
     @property
     def host(self) -> H:
         """Host associated with this path."""
         return self._host
 
+    def _local_str(self) -> str:
+        return str(self._path)
+
+    def at_host(self, host: H) -> str:
+        """
+        Convert this ``Path`` into a string representation assuming it should
+        be valid for the machine ``host``.  An exception is raised if this
+        ``Path`` is for a different machine instead.
+
+        This prevents accidentally using a ``Path`` with the wrong host as much
+        as possible.
+        """
+        if self.host != host:
+            raise Exception(f"This path ({self}) is not for the host {host!r}")
+        return str(self._path)
+
+    # }}}
+
+    # PurePosixPath like API {{{
+    # Mimick the API of pathlib's PurePosixPath as much as possible.  This
+    # allows users to use tbot's Path just like the one they know from pathlib.
+    def _prepare_args_list(self, args: Iterable[Any]) -> List[Any]:
+        ret = list()
+        for arg in args:
+            if isinstance(arg, Path):
+                assert arg.host == self.host
+                ret.append(arg._path)
+            else:
+                ret.append(arg)
+        return ret
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._host == other._host and self._path == other._path
+
+    def __hash__(self) -> int:
+        return hash((self._host, self._path))
+
+    def __lt__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path < other._path
+
+    def __le__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path <= other._path
+
+    def __gt__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path > other._path
+
+    def __ge__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path >= other._path
+
+    @property
+    def name(self) -> str:
+        return self._path.name
+
+    @property
+    def suffix(self) -> str:
+        return self._path.suffix
+
+    @property
+    def suffixes(self) -> List[str]:
+        return self._path.suffixes
+
+    @property
+    def stem(self) -> str:
+        return self._path.stem
+
+    def with_name(self, name: str) -> "Path[H]":
+        return Path(self._host, self._path.with_name(name))
+
+    def with_stem(self, stem: str) -> "Path[H]":
+        return Path(self._host, self._path.with_name(stem + self._path.suffix))
+
+    def with_suffix(self, suffix: str) -> "Path[H]":
+        return Path(self._host, self._path.with_suffix(suffix))
+
+    def relative_to(self, *other: Any) -> "Path[H]":
```

Comment:
Should this not enforce `other` to be a tbot Path? If it is, I see that `_prepare_args_list` throws an error if `self` and `other` are not associated with the same host, good. But was it your intention to allow strings and Python paths as args? Hm, I guess it's fine as ValueError will be raised by `self._path.relative_to` if the paths are not even relative to each other?

Reply:
My initial implementation was enforcing a path, but it turns out the method (in `pathlib`) was conceived to allow multiple path segments as well:

```python
x = pathlib.Path("/tmp/foo/bar")
x.relative_to("/tmp", "foo")
```

I think it would be rather user-unfriendly if all those segements needed to be tbot-paths.  So my design decision was this:  Allow strings (and pathlib paths) as well, but do check any tbot-paths which are passed in.  Additionally, the returned `Path` is always associated with the same host as the original, so even if someone does something nasty, it will fail further down the line.  And, as you noticed, if the Path isn't even a subpath, we'll get a `ValueError` anyway.